squigglepy/correlation.py

```diff
@@ -0,0 +1,237 @@
+#   list of conditions and the following disclaimer.
+
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+import numpy as np
+from scipy.linalg import cholesky
+from scipy.stats import rankdata
+from scipy.stats.distributions import norm
+from numpy.typing import NDArray
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from .distributions import OperableDistribution
+
+
+def correlate(
+    variables: tuple[OperableDistribution, ...],
+    correlation: NDArray[np.float64] | list[list[float]] | np.float64 | float,
+):
+    """
+    Correlate a set of variables according to a rank correlation matrix.
+
+    This employs the Iman-Conover method to induce the correlation while
+    preserving the original marginal distributions.
+
+    Parameters
+    ----------
+    variables : tuple of distributions
+        The variables to correlate as a tuple of distributions.
+
+    correlation : 2d-array or float
+        An n-by-n array that defines the desired Spearman rank correlation coefficients.
+        This matrix must be symmetric and positives emi-definite; and must not be confused with
+        a covariance matrix.
+
+        Correlation parameters can only be between -1 and 1, exclusive
+        (including extremely close approximations).
+
+        If a float is provided, all variables will be correlated with the same coefficient.
+
+    Returns
+    -------
+    correlated_variables : tuple of distributions
+        The correlated variables as a tuple of distributions in the same order as
+        the input variables.
+
+    Examples
+    --------
+    >>> a, b = sq.uniform(-1, 1), sq.to(0, 3)
+    >>> a, b = sq.correlate((a, b), [[1, 0.9], [0.9, 1]])
```

Comment:
I'm guessing we want the default interface to be `sq.correlate([a, b], 0.9)` and the `[[1, 0.9], [0.9, 1]]` can then be inferred? This seems cleaner. Though perhaps this list could still be passed anyway if desired?

Reply:
This has already been implemented! I just haven't added an example for it.

Reply:
(As a rule of thumb, don't expect draft PRs to be in a coherent state. I usually don't expect people to review the code before I mark it as ready for review.)